sound/soc/sof/amd/acp.c

```diff
@@ -477,6 +477,86 @@ int amd_sof_acp_resume(struct snd_sof_dev *sdev)
 }
 EXPORT_SYMBOL_NS(amd_sof_acp_resume, SND_SOC_SOF_AMD_COMMON);
 
+#if IS_ENABLED(CONFIG_SND_SOC_SOF_AMD_SOUNDWIRE)
+static int acp_sof_scan_sdw_devices(struct snd_sof_dev *sdev, u64 addr)
+{
+	struct acpi_device *sdw_dev;
+	struct acp_dev_data *acp_data;
+	const struct sof_amd_acp_desc *desc = get_chip_info(sdev->pdata);
+
+	if (!addr)
+		return -ENODEV;
+
+	acp_data = sdev->pdata->hw_pdata;
+	sdw_dev = acpi_find_child_device(ACPI_COMPANION(sdev->dev), addr, 0);
+	if (!sdw_dev)
+		return -ENODEV;
+
+	acp_data->info.handle = sdw_dev->handle;
+	acp_data->info.count = desc->sdw_max_link_count;
+
+	return amd_sdw_scan_controller(&acp_data->info);
+}
+
+static int amd_sof_sdw_probe(struct snd_sof_dev *sdev)
+{
+	struct acp_dev_data *acp_data;
+	struct sdw_amd_res sdw_res;
+	void *sdw;
+
+	acp_data = sdev->pdata->hw_pdata;
+
+	memset(&sdw_res, 0, sizeof(sdw_res));
+	sdw_res.addr = acp_data->addr;
+	sdw_res.reg_range = acp_data->reg_range;
+	sdw_res.handle = acp_data->info.handle;
+	sdw_res.parent = sdev->dev;
+	sdw_res.dev = sdev->dev;
+	sdw_res.acp_lock = &acp_data->acp_lock;
+	sdw_res.count = acp_data->info.count;
+	sdw_res.link_mask = acp_data->info.link_mask;
+	sdw_res.mmio_base = sdev->bar[ACP_DSP_BAR];
+
+	sdw = sdw_amd_probe(&sdw_res);
+	if (!sdw) {
+		dev_err(sdev->dev, "SoundWire probe failed\n");
+		return -EINVAL;
+	}
+
+	/* save context */
+	acp_data->sdw = sdw;
+	return sdw_amd_startup(acp_data->sdw);
```

Comment:
so why do you need to implement the startup callback? Everything can be moved to `sdw_amd_probe(&sdw_res);`

Reply:
Sorry, not following any of this.

````c
	sdw = sdw_amd_probe(&sdw_res);
	if (!sdw) {
		dev_err(sdev->dev, "SoundWire probe failed\n");
		return -EINVAL;
	}

	/* save context */
	acp_data->sdw = sdw;
	return sdw_amd_startup(acp_data->sdw);
````

why can't you pass acp_data in the probe and store the context inside of sdw_amd_probe? 

You are trying to stick to an Intel definition of interfaces which makes no sense for you.

Reply:
1. acp_data is not common private data structure for legacy stack and sof stack. It will be different one for legacy stack.
2. We want to abstract usage of parent driver structures in amd_init.c file.
3. I think its still possible, passing parent driver private data structure sdw member to the probe function and assign it in probe function itself followed by invoking start sequence in the same probe call.
4. We want to divide whole probe and start functionality in two functions to increase the readability and scalability and we want to leave a place holder so that in future if start() functionality needs to move to some other place for example after FW loading, there will no option left with this approach. 
5. For example, we need to support a SOF scenario. In case of firmware load failure, there is no point in starting the soundwire managers. In this case if we don't have startup sequence separately, we can't support it.

We don't see any major issue in exporting one more function(start() function) other than probe().
Please note our intention is not to replicate everything in Intel style.

Reply:
You really want to start the links as early as possible to make sure the peripherals are enumerated in parallel with the rest of firmware loading.

I can't force you to listen or change your code, but at this point you really don't have a good technical rationale for splitting probe/start. I am not the final reviewer so others might have a different opinion.

Reply:
@plbossart : We accept your review comments. We will do the code modifications and respin the patch set and let you know the test results.

Reply:
I agree with @plbossart.  `sdw_amd_startup()` is only used here. It is not necessary to have a separated startup function.

Reply:
@bardliao : will make the changes as per @plbossart comments will combine probe and start sequence and remove separate export function.